include/runtime/collections/RBTree.h

```diff
@@ -0,0 +1,512 @@
+#ifndef RBTREE_HEADER_H
+#define RBTREE_HEADER_H
+
+/* Adapted from https://github.com/BartoszMilewski/Okasaki/tree/master/RBTree */
+/* Implementing algorithms from                                               *
+ * https://matt.might.net/papers/germane2014deletion.pdf                      */
+
+#include <cassert>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+#define CONSTRUCT_MSG_AND_THROW(msg)                                           \
+  do {                                                                         \
+    std::string emsg = std::string("[") + std::string(__func__)                \
+                       + std::string("] ") + std::string(msg);                 \
+    throw std::invalid_argument(emsg);                                         \
+  } while (false)
+
+//---               Ordered map on top of a red-black tree.                ---//
+
+// 1. No red node has a red child.
+// 2. Every path from root to empty node contains the same
+// number of black nodes.
+
+template <class T, class V>
+class RBTree {
+  enum Color { R, B, BB };
+
+  static Color minusOneColor(Color c) {
+    switch (c) {
+    case B: return R;
+    case BB: return B;
+    default: CONSTRUCT_MSG_AND_THROW("Unexpected color");
+    }
+  }
+
+  struct Node {
+    Node(Color c)
+        : _c(c)
+        , _s(0) { }
+    Color _c;
+    size_t _s;
+
+    virtual bool isLeaf() const = 0;
+    virtual ~Node(){};
+  };
+
+  struct Leaf : public Node {
+    Leaf(Color c)
+        : Node(c) {
+      assert(c == B || c == BB);
+    }
+    virtual bool isLeaf() const override { return true; }
+    virtual ~Leaf(){};
+  };
+
+  struct InternalNode : public Node {
+    InternalNode(
+        Color c, std::shared_ptr<const Node> const &lft, T key, V val,
+        std::shared_ptr<const Node> const &rgt)
```

Comment:
Given that the constructor immediately constructs the member variables `_lft` and `_rgt` with these const-ref arguments, they should just be taken by value in the first place. This helps the compiler figure out when to make copies a bit better than copying from a const-ref. This repeats elsewhere.

Reply:
I changed this in the constructors where a copy would immediately follow. There is one case left where this is not the case, rather members of the arguments are given to the constructor that then performs the actual copies.